## Problem 6: Sum Square Difference

The sum of the squares of the first ten natural numbers is,

$1^2 + 2^2 + \ldots + 10^2 = 385$

The square of the sum of the first ten natural numbers is,

$(1 + 2 + \ldots + 10)^2 = 3025$

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is $3025 - 385 = 2640$.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [1]:
''' 
Calculates the sum of the squares of the integers up to and including n_max.

eg:
>>> sum_of_square(10)
>>> 385
'''

def sum_of_square(n_max : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the sum of the squares of the integers up to and including n_max.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : n_max : The maximum number to include in the sum.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> sum_of_square(10)
        >>> 385
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(n_max, int): raise ValueError('Please enter an integer > 0 for the n_max argument.')
    if n_max < 1: raise ValueError('Please enter an integer > 0 for the n_max argument.')
    
    # Loop through all integers from 1 to n_max, square them, then add them to the sum.
    square_sum = 0
    for n in range(1, n_max+1):
        square_sum += n**2
        
    return(square_sum)

In [2]:
''' 
Calculates the square of the sum of the integers up to and including n_max.

eg:
>>> square_of_sum(10)
>>> 3025
'''

def square_of_sum(n_max : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the square of the sum of the integers up to and including n_max.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : n_max : The maximum number to include in the sum.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> square_of_sum(10)
        >>> 3025
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(n_max, int): raise ValueError('Please enter an integer > 0 for the n_max argument.')
    if n_max < 1: raise ValueError('Please enter an integer > 0 for the n_max argument.')
    
    # Loop through all integers from 1 to n_max, sum them, then square the sum.
    sum_square = 0
    for n in range(1, n_max+1):
        sum_square += n
    
    return(sum_square**2)

In [3]:
''' 
Calculates the difference between the sum of squares and square of sum.

eg:
>>> square_difference(10)
>>> 2640
'''

def square_difference(n_max):
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the difference between the sum of squares and square of sum.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : n_max : The maximum number to include in the calculation.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> square_difference(10)
        >>> 2640
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(n_max, int): raise ValueError('Please enter an integer > 0 for the n_max argument.')
    if n_max < 1: raise ValueError('Please enter an integer > 0 for the n_max argument.')
    
    square_sum = sum_of_square(n_max)
    sum_square = square_of_sum(n_max)
    
    return(sum_square - square_sum)

In [4]:
max_n = 100
solution = square_difference(max_n)
print(f'The difference between the sum of the squares of the first {max_n} natural numbers and the square of the sum = {solution:,}.')

The difference between the sum of the squares of the first 100 natural numbers and the square of the sum = 25,164,150.


There are also analytical formulas for these quantities:

The sum of natural numbers is given by $s = N(N + 1) / 2$.

The sum of the squares of the natural numbers is given by $s_2 = N(N + 1)(2N + 1) / 6$.

Therefore $s^2 - s_2 = (N^4 / 4) + (N^3 / 6) - (N^2 / 4) - (N / 6)$.

In [5]:
''' 
Calculates the difference between the sum of squares and square of sum using analytical formulae:

s = N(N + 1) / 2
s_2 = N(N + 1)(2N + 1) / 6

s^2 - s_2 = (N^4 / 4) + (N^3 / 6) - (N^2 / 4) - (N / 6)

eg:
>>> analytical_square_difference(10)
>>> 2640
'''

def analytical_square_difference(n_max):
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the difference between the sum of squares and square of sum using analytical formulae:

        s = N(N + 1) / 2
        s_2 = N(N + 1)(2N + 1) / 6

        s^2 - s_2 = (N^4 / 4) + (N^3 / 6) - (N^2 / 4) - (N / 6)
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : n_max : The maximum number to include in the calculation.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> analytical_square_difference(10)
        >>> 2640
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(n_max, int): raise ValueError('Please enter an integer > 0 for the n_max argument.')
    if n_max < 1: raise ValueError('Please enter an integer > 0 for the n_max argument.')
    
    return int(((n_max**4)/4) + ((n_max**3)/6) - ((n_max**2)/4) - (n_max/6))

In [6]:
max_n = 100
solution = analytical_square_difference(max_n)
print(f'The difference between the sum of the squares of the first {max_n} natural numbers and the square of the sum = {solution:,}.')

The difference between the sum of the squares of the first 100 natural numbers and the square of the sum = 25,164,150.


### Problem 6 Solution: 

The difference between the sum of the squares of the first 100 natural numbers and the square of the sum = 25,164,150.